### PR TITLE
*: switch to BusLocator-oriented helpers

### PR DESCRIPTION
**Pull request by vcaputo**
Mostly mechanical substitutions reducing some verbosity

### CODE REVIEW

**Comment by vcaputo** [comment]
Did more of the BusLocator spring cleaning during some down time, posted as draft until passes CI.

**Comment by DaanDeMeyer** [comment]
Only a small nit, but you can declare the locator constants as follows:

```c
static const BusLocator *home_mgr = &(BusLocator) {
        .destination = "org.freedesktop.home1",
        .path = "/org/freedesktop/home1",
        .interface = "org.freedesktop.home1.Manager",
};
```

and then you don't have to use the address operator each time you're passing it to a function:

```c
r = bus_call_method(bus, home_mgr, "ListHomes", &error, &reply, NULL);
```

**Comment by vcaputo** [comment]
True, maybe in a subsequent commit that can be tweaked along with the existing examples in src/home?

I feel kind of partial to the &foo_mgr syntax since it also serves as syntactic sugar clearly expressing this is a pointer at the call site though.

**Comment by DaanDeMeyer** [comment]
I'd argue using the address operator explicitly at the call site is especially useful when the function either modifies the pointer's value or stores the pointer somewhere. In other cases, we don't really care whether we're passing by value or by pointer so I don't think there's a lot of added value in making it explicit that we're passing a pointer at the call site in this case.

I can make a follow up PR that changes this everywhere (including home) if you'd rather not add it to this PR.

**Comment by vcaputo** [comment]
I don't feel strongly either way, how about we get another opinion from one of the other core folks like @keszybz or @poettering on what color to paint this particular bikeshed

I'm not averse to updating this PR to reflect whatever is the prevailing preference, it is a draft after all.

File: src/locale/localectl.c [hunk 1]
```diff
@@ -163,13 +169,7 @@ static int set_locale(int argc, char **argv, void *userdata) {
...
         polkit_agent_open_if_enabled(arg_transport, arg_ask_password);
 
-        r = sd_bus_message_new_method_call(
-                        bus,
-                        &m,
-                        "org.freedesktop.locale1",
-                        "/org/freedesktop/locale1",
-                        "org.freedesktop.locale1",
-                        "SetLocale");
+        r = bus_message_new_method_call( bus, &m, &locale_mgr, "SetLocale");
```
Comment (keszybz) [inline src/locale/localectl.c]: Spurious " " here.

File: src/locale/localectl.c [hunk 2]
```diff
@@ -215,14 +215,7 @@ static int set_vconsole_keymap(int argc, char **argv, void *userdata) {
...
 
-        r = sd_bus_call_method(
-                        bus,
-                        "org.freedesktop.locale1",
-                        "/org/freedesktop/locale1",
-                        "org.freedesktop.locale1",
-                        "SetVConsoleKeyboard",
-                        &error,
-                        NULL,
+        r = bus_call_method( bus, &locale_mgr, "SetVConsoleKeyboard", &error, NULL,
```
Comment (keszybz) [inline src/locale/localectl.c]: Here too.

File: src/locale/localectl.c [hunk 3]
```diff
@@ -258,14 +251,7 @@ static int set_x11_keymap(int argc, char **argv, void *userdata) {
...
 
-        r = sd_bus_call_method(
-                        bus,
-                        "org.freedesktop.locale1",
-                        "/org/freedesktop/locale1",
-                        "org.freedesktop.locale1",
-                        "SetX11Keyboard",
-                        &error,
-                        NULL,
+        r = bus_call_method(bus, &locale_mgr, "SetX11Keyboard", &error, NULL,
```
Comment (keszybz) [inline src/locale/localectl.c]: Indentation.

File: src/login/loginctl.c [hunk 4]
```diff
@@ -1017,11 +988,7 @@ static int activate(int argc, char *argv[], void *userdata) {
 
         for (i = 1; i < argc; i++) {
 
-                r = sd_bus_call_method(
-                                bus,
-                                "org.freedesktop.login1",
-                                "/org/freedesktop/login1",
-                                "org.freedesktop.login1.Manager",
+                r = bus_call_method(bus, &login_mgr,
```
Comment (keszybz) [inline src/login/loginctl.c]: It'd be nicer to move "bus, &login_mgr," to a new line and indent it the same as the following arguments.

File: src/login/loginctl.c [hunk 5]
```diff
@@ -1050,13 +1017,7 @@ static int kill_session(int argc, char *argv[], void *userdata) {
...
         for (i = 1; i < argc; i++) {
 
-                r = sd_bus_call_method(
-                        bus,
-                        "org.freedesktop.login1",
-                        "/org/freedesktop/login1",
-                        "org.freedesktop.login1.Manager",
-                        "KillSession",
-                        &error, NULL,
+                r = bus_call_method(bus, &login_mgr, "KillSession", &error, NULL,
```
Comment (keszybz) [inline src/login/loginctl.c]: Same here and below.

File: src/machine/machined-dbus.c [hunk 6]
```diff
@@ -31,6 +31,12 @@
 #include "unit-name.h"
 #include "user-util.h"
 
+static const BusLocator systemd_mgr = {
+        .destination = "org.freedesktop.systemd1",
+        .path = "/org/freedesktop/systemd1",
+        .interface = "org.freedesktop.systemd1.Manager"
+};
```
Comment (keszybz) [inline src/machine/machined-dbus.c]: Shouldn't this be declared once in src/shared/ somewhere?

**Comment by vcaputo** [comment]
@keszybz I believe I've addressed all of the requested changes, promoted from draft to PR, PTAL

File: src/shared/bus-locator.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+extern const BusLocator *bus_home_mgr;
+extern const BusLocator *bus_import_mgr;
+extern const BusLocator *bus_locale;
+extern const BusLocator *bus_login_mgr;
+extern const BusLocator *bus_machine_mgr;
+extern const BusLocator *bus_network_mgr;
+extern const BusLocator *bus_portable_mgr;
+extern const BusLocator *bus_resolve_mgr;
+extern const BusLocator *bus_systemd_mgr;
+extern const BusLocator *bus_timedate;
```
Comment (keszybz) [inline src/shared/bus-locator.h]: This should be `const BusLocator* const bus_...;`. Right now the pointers themselves are mutable.

**Comment by keszybz** [comment]
I think this should be followed by a similar PR to hide `&` from the BusImplementation structures...

File: src/shared/bus-locator.h (hunk 7 above)
Comment (vcaputo) [reply src/shared/bus-locator.h]: Duh, fixed.

**Comment by vcaputo** [comment]
Since this had the good-to-merge label added and I addressed the minor suggestion, I went ahead and merged to avoid the high probability of needing a rebase being such a sprawling tree-wide PR

**Comment by vcaputo** [comment]
thanks for the reviews

**Comment by vcaputo** [comment]
> I think this should be followed by a similar PR to hide `&` from the BusImplementation structures...

@keszybz I made a quick stab at this but was met with an annoying compilation error that quickly became more trouble than I can spend on this right now...  Here's the branch of where I left things in case you want to pick it up and make the compiler happy then finish whatever's left to do.

 https://github.com/vcaputo/systemd/tree/boip

The error I got bogged down at is:
```
In file included from ../src/login/logind-dbus.c:17:0:
../src/shared/bus-util.h:45:71: error: initializer element is not constant
 #define BUS_IMPLEMENTATIONS(...) ((const BusObjectImplementation* []) { __VA_ARGS__, NULL })
                                                                       ^
../src/login/logind-dbus.c:3759:21: note: in expansion of macro ‘BUS_IMPLEMENTATIONS’
         .children = BUS_IMPLEMENTATIONS(seat_object,
                     ^~~~~~~~~~~~~~~~~~~
../src/shared/bus-util.h:45:71: note: (near initialization for ‘(anonymous)’)
 #define BUS_IMPLEMENTATIONS(...) ((const BusObjectImplementation* []) { __VA_ARGS__, NULL })
                                                                       ^
../src/login/logind-dbus.c:3759:21: note: in expansion of macro ‘BUS_IMPLEMENTATIONS’
         .children = BUS_IMPLEMENTATIONS(seat_object,
                     ^~~~~~~~~~~~~~~~~~~
```

**Pull request merged**